src/simulation/LoadGenerator.cpp

```diff
@@ -915,6 +927,46 @@ LoadGenerator::getNextAvailableAccount()
     uint64_t sourceAccountId = *it;
     mAccountsAvailable.erase(it);
     releaseAssert(mAccountsInUse.insert(sourceAccountId).second);
+
+    if (mApp.getHerder().sourceAccountPending(
+            findAccount(sourceAccountId, ledgerNum)->getPublicKey()))
+    {
+        // Although mAccountsAvailable shouldn't contain pending accounts, it is
+        // possible when the network is overloaded. Consider the following
+        // scenario:
+        // 1. This node generates a transaction `t` using account `a` and
+        //    broadcasts it on. In doing so, loadgen marks `a` as in use,
+        //    removing it from `mAccountsAvailable.
+        // 2. For whatever reason, `t` never makes it out of the queue and this
+        //    node bans it.
+        // 3. After some period of time, this node unbans `t` because bans only
+        //    last for so many ledgers.
+        // 4. Loadgen marks `a` available, moving it back into
+        //    `mAccountsAvailable`.
+        // 5. This node hears about `t` again on the network and (as it is no
+        //    longer banned) adds it back to the queue
+        // 6. getNextAvailableAccount draws `a` from `mAccountsAvailable`.
+        //    However, `a` is no longer available as `t` is in the transaction
+        //    queue!
+        //
+        // In this scenario, returning `a` results in an assertion failure
+        // later. At this point we have two reasonable options:
+        // 1. Mark `a` in-use and resample new accounts until we either find one
+        //    that is available, or run out of accounts
+        // 2. Fail loadgen.
+        // We chose option (2), as this scenario is indicative of an overloaded
+        // network and in practice loadgen is highly likely to fail soon anyway.
+        // Moreover, option (1) provides some false sense of having resolved the
+        // issue as this is a race condition that could still occur between
+        // returning the new account from this function and using it later on.
+        CLOG_WARNING(
+            LoadGen,
+            "mAccountsAvailable contains an account '{}' already in use",
+            sourceAccountId);
+        mFailed = true;
```

Comment:
We could also just pick up another account to use to avoid failing loadgen? This edge case seems quite rare and hard to trigger, so I think we should try to keep loadgen alive if possible (I don't think we can draw conclusions on network being overloaded based on transactions re-surfacing)

Reply:
(this way getNextAvailableAccount would always return something instead of an optional, so we won't need to check for nullopt in other places)

Reply:
I just pushed a change that resamples instead, and dropped the `nullopt`